### PR TITLE
Enhances Spark job to account for single-sided spans

### PR DESCRIPTION
**Pull request by yurishkuro**
where either parent or child is not instrumented.

Addresses https://github.com/openzipkin/zipkin/issues/914 and https://github.com/openzipkin/zipkin/issues/917.

The following steps are:
* change test in https://github.com/openzipkin/zipkin/issues/917, possibly `@Ignore` in stores that cannot support this yet
* move the new tests to Zipkin, possibly `@Ignore` in stores that cannot support this yet
* extend common data model to allow `instrumentation` field on `DependencyLink`, add more tests
* update this repo to support new tests with the extra field
* extend JSON API to return the extra field

### CODE REVIEW

**Comment by yurishkuro** [comment]
cc @kristofa @adriancole @jcarres-mdsol

**Comment by adriancole** [comment]
Interesting.. I like how you are experimenting without breaking core ;) although we clearly need to follow-up if this is to be widely supported.

Can you put into the description the api-level change? Like adds Span.blah which is an enum of foo, bar baz, describing each choice. This will help introduce the effort, and help us think if through without reading scala ;)

Cc @t-yuki

**Comment by yurishkuro** [comment]
@adriancole currently there is no API change. The job internally keeps track of the inferred instrumentation level of each edge, but discards it before saving to the store. In theory this diff can be merged by itself and give people richer dependencies diagram, without affecting core (but at the cost of overriding one test, unless we change it per https://github.com/openzipkin/zipkin/issues/917#issuecomment-176764673).

File: cassandra/src/main/scala/io/zipkin/dependencies/spark/cassandra/ZipkinDependenciesJob.scala [hunk 1]
```diff
@@ -111,28 +221,36 @@ case class ZipkinDependenciesJob(sparkMaster: String = ZipkinDependenciesJob.spa
...
-      println(s"NO VALID SPANS FOUND IN THE TIME RANGE")
+      println(s"NO LINKS WERE GENERATED FROM SPANS")
     } else {
       val dependencies: DependenciesInfo = aggregates
-        .map { case ((parent: String, child: String), callCount: Long) =>
-        DependenciesInfo(Seq(DependencyLinkInfo(parent = parent, child = child, callCount = callCount)))
-      }
-        .reduce(_ + _) // merge DLs under one Dependencies object, which overrides +
+        .map { case ((parent: String, child: String), link: Link) =>
+        // TODO signal loss here - `instrumentation` is ignored
```
Comment (adriancole) [inline cassandra/src/main/scala/io/zipkin/dependencies/spark/cassandra/ZipkinDependenciesJob.scala]: ok so here's where you mean that there's no api impact, as the intermediate state is thrown out

**Comment by jcarres-mdsol** [comment]
A comment not for this PR but for the upcoming. The more unique functionality different stores have the more difficult maintenance will become. At some point it would not even be clear what works on each.

**Comment by adriancole** [comment]
I'm primarily concerned about the complexity added here. I have re-read this a few times, and thanks to your comments I think I understand what's going on. 

First reaction to this is that we are adding a lot of code that is nuanced, particularly around name resolution, priority, etc. A lot of this work is around when instrumentation log different names than servers, and cannot be corrected. I'm sure some deployments cannot correct service names, but it seems interesting to only correct this problem in the dependency graph.

IOTW, If it isn't tenable to have instrumentation log correct service names, (or log empty if they don't know), maybe a data cleanup job can be done somewhere else. Reason is that bad service names apply equally here and anything else including api queries.

I recall groupon saying they had a classification job as a part of their collector pipeline, for example. This is an unsurprising approach. When I was at twitter, we didn't use Spark, but there were several zipkin hadoop jobs, some of them were data cleaning. IOTW it wasn't one job that did everything including creating the dependency graph.

**Comment by yurishkuro** [comment]
@jcarres-mdsol see Capabilities API to address differences in stores https://github.com/openzipkin/zipkin/issues/931. I don't think inability of certain backends to support features of other backends should be a blocker for developing those new features. That would only encourage people to make private forks of Zipkin.

@adriancole I agree in principle that data cleansing jobs can be separate, however there is no data cleansing happening here. The existing job was already inferring the other not-instrumented service when the client is instrumented, only it did it by accident. Client creates a separate span for an rpc and the SA takes precedence on the serviceName, so the span fully emitted by the client is counted as "child" in the Diagram.  My change extends this accidental & implicit logic into explicit one that works for the opposite case as well, where server is instrumented but the client is not. The current job would not build a link for that case, even though the data is there - I consider it a bug.

Note that the logic would've been much simpler if we enforced two spans per RPC.

**Comment by yurishkuro** [comment]
FYI - I want to see if there is a way to achieve the same results with less data stored in the transient span

**Comment by adriancole** [comment]
I think if this change's description had less scope, ex didn't have a specific change to roll a specific instrumented field all the way including the json v1 api, I'd be far more in favor of it.

Ex a PR who uses tools like transient intermediate state to fix missing uninstrumented server links is more than reasonable.

FWIW, I also want to understand how we can single-side spans more generically, but I would rather place that energy into making single-host spans *a requirement* of v2, then use the v2 api in the aggregation system.. this solves a lot of complexity and doesn't require us to backport (and lose even more time) in v1.

**Comment by adriancole** [comment]
PS having spent a lot of time in the aggregation code lately, I think this particular request is helpful as a bridge.

For example, if we ack that single-host spans exist, and might be default in v2, we can practice with the aggregation code (by making an intermediate span type for example) as the aggregation model is far simpler (even if parsing isn't)

Long story short.. Let's finish this change. I think there's only one test here not upstream at the moment. I can look into that.

**Comment by yurishkuro** [comment]
in the current form (even after removing the new `instrumentation` field), this job takes a bit more memory and I was not able to run it to completion on our test cluster to confirm that it actually makes a difference in the real dependency diagram. That's why I haven't spent more time on it, waiting to migrate to a bigger cluster. I should be able to return to it closer to the end of next week.

**Comment by adriancole** [comment]
Thanks for the update!

**Comment by adriancole** [comment]
@yurishkuro any update on this?

**Comment by yurishkuro** [comment]
nothing good, sorry. Our hadoop cluster required that I downgrade the job to Java7, which effectively prevented me from extending this OSS code.

**Comment by adriancole** [comment]
all tests pass now (and the job compiles to java 7 bytecode) per #29

**Pull request closed without merging**